1) What is the meaning of hyper parameter runung
     Is the process Evaluate the model performance by changing parameters
2) What is the meaning of  GridSearchCV and Random Search CV(cross validation)
    Checking the Model performance by applying
    All combinations of hyper parameter one by one

     Checking the Model performance by applying
     random combinations of hyper parameter one by one
3) What is the meaning of  CV
        CV=4 randomly divide in 4 parts
        p1,p2,p3,p4
       Model1:Train data:P1+P2+P3 Test Data:p4  acc1(accuracy)
       Model1:Train data:P1+P2+P4 Test Data:p1  acc2
       Model1:Train data:P1+P3+P4 Test Data:p3  acc3
       Model1:Train data:P2+P3+P4 Test Data:p4  acc4
       avg of (a1+a2+a3+a4)/4
4) What is the procedure to get Best parameters
       avg of (a1+a2+a3+a4)/4  for all parameters
       the model wiht a higest accuracy is the higest arrucy
        

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme(color_codes=True)  # to get different colors
pd.set_option('display.max_columns',None)  # To display the max clumns
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score,classification_report,roc_auc_score,roc_curve,auc
from sklearn.metrics import confusion_matrix,ConfusionMatrixDisplay


## $step-2$

## Read the data


In [6]:
data=pd.read_csv(r"C:\Data Sets\preprocessd_data.csv")
data

gender       age  no_of_days_subscribed  multi_screen  mail_subscribed  \
0          0 -0.264643              -0.946659             0                0   
1          0  0.029654               1.235009             0                0   
2          0  2.580235               0.658246             0                0   
3          0 -1.441834               0.783629             0                1   
4          0  0.127754               2.288228             0                0   
...      ...       ...                    ...           ...              ...   
3401       1 -0.805656               0.479710             0                0   
3402       0  0.477826               0.674425             0                0   
3403       0 -0.550521               0.426099             0                1   
3404       0  0.004651              -0.444637             0                0   
3405       0 -0.057028               0.317748             0                0   

      weekly_mins_watched  minimum_daily_mins  weekly_max_night_mins  \
0               -1.508785            0.722258              -0.944404   
1                0.304821           -0.901748              -0.688523   
2               -2.266627            0.613991              -0.483818   
3                0.638123           -0.252146               0.079121   
4               -0.333851            0.253101              -0.893228   
...                   ...                 ...                    ...   
3401             0.887802            0.552184               0.865765   
3402             2.038061           -0.931071               0.567257   
3403            -0.410669            0.344847              -0.729485   
3404             1.359751           -0.847121              -0.715010   
3405             2.844852            0.098115              -0.550251   

      videos_watched  customer_support_calls  churn  
0          -1.403000               -0.415473      0  
1          -0.599124                0.347522      0  
2          -1.403000                2.636507      1  
3          -0.197185                1.110517      0  
4           1.008630               -0.415473      0  
...              ...                     ...    ...  
3401       -0.495634               -0.415473      1  
3402        2.100185                0.235227      1  
3403       -0.403904                1.873512      1  
3404       -0.197185               -1.178468      1  
3405        0.606692               -1.178468      1  

[3406 rows x 11 columns]

## divide the data into input and output

In [8]:
x=data.drop('churn',axis=1)
y=data['churn']

## Train test split

In [10]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(2724, 10)
(682, 10)
(2724,)
(682,)


## $Step-5$

## Read the base model

In [12]:
from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.tree import DecisionTreeClassifier
grid_tree=DecisionTreeClassifier() #base model
grid_tree

# we are not fitting the data

DecisionTreeClassifier()

## Creating the paramater file

In [13]:
grid_tree.get_params()

{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'random_state': None,
 'splitter': 'best'}

In [14]:
# You need to create dictonary with hyper parametets

param_grid={
    'criterion':['gini','entropy'],#2
    'max_depth':[3,4,5,6,7,8], #6
    'min_samples_split':[2,3,4], #3
    'min_samples_leaf':[1,2,3,4], #4
    'random_state':[0,42] #2
}
# 2*6*3*4*2=288

## Step-7

In [16]:
# now these parameters to apply in a DT model

grid_search= GridSearchCV(grid_tree, # Base model
                          param_grid, #Params
                         scoring='accuracy', #metric
                         cv=5,
                         verbose=True)
#Verbose=True I will show in deep learning indetail

In [17]:
grid_search

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 4, 5, 6, 7, 8],
                         'min_samples_leaf': [1, 2, 3, 4],
                         'min_samples_split': [2, 3, 4],
                         'random_state': [0, 42]},
             scoring='accuracy', verbose=True)

## $step-8$

## Fit the data

In [19]:
import time
start=time.time()
grid_search.fit(x_train,y_train)
#generally we will padd the entire data
#x_train=input data
#y_train=outpur data
end=time.time()
print('Total timme take is:',(end-start))

Fitting 5 folds for each of 288 candidates, totalling 1440 fits
Total timme take is: 16.668433904647827


C:\anaconda\Lib\site-packages\numpy\ma\core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


In [20]:
dir(grid_search)

['__abstractmethods__',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__sklearn_clone__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_build_request_for_signature',
 '_check_feature_names',
 '_check_n_features',
 '_check_refit_for_multimetric',
 '_doc_link_module',
 '_doc_link_template',
 '_doc_link_url_param_generator',
 '_estimator_type',
 '_format_results',
 '_get_default_requests',
 '_get_doc_link',
 '_get_metadata_request',
 '_get_param_names',
 '_get_routed_params_for_fit',
 '_get_scorers',
 '_get_tags',
 '_more_tags',
 '_parameter_constraints',
 '_repr_html_',
 '_repr_html_inner',
 '_repr_mimebundle_',
 '_required_parameters',
 '_run

**Estimator means model**

- after developed model 1440 times

    - 288 paramters comintation with CV of 5

    - Each patamaters combination trained 5 times

    - Each paramater combination will get 5 accuracy values

    - Each parameter combinen will get one Avg acuracy

    - Like that we will get 288 Avg accuracy values

    - In that one paramater combination has highest avetage accuracy

    -  In that one parameter combineation has higest avg accuracy

    - That is the best patameter
 
    - the model is  using best patameters called as best estimator

In [22]:
grid_search.best_estimator_ #estimator means modle

DecisionTreeClassifier(criterion='entropy', max_depth=8, min_samples_leaf=4,
                       random_state=0)

In [27]:
 grid_search.cv_results_

{'mean_fit_time': array([0.01354008, 0.01126022, 0.00713181, 0.00713706, 0.00716391,
        0.0069006 , 0.00577641, 0.00550451, 0.00593104, 0.00588355,
        0.00589223, 0.00556321, 0.00629234, 0.0065227 , 0.0059195 ,
        0.00614581, 0.00641241, 0.00642576, 0.00698395, 0.0069798 ,
        0.0063765 , 0.00604153, 0.00592065, 0.0057878 , 0.00705085,
        0.00647144, 0.00658169, 0.00659571, 0.00699606, 0.00668383,
        0.00678134, 0.00690289, 0.00661511, 0.00681596, 0.00681438,
        0.00711279, 0.00748467, 0.00632324, 0.00646567, 0.00706177,
        0.00693302, 0.00712109, 0.00747771, 0.00650849, 0.00772371,
        0.00731835, 0.00661278, 0.00679431, 0.00792131, 0.00812564,
        0.00789814, 0.00811958, 0.00801764, 0.00787382, 0.00790553,
        0.00830774, 0.00800433, 0.00815668, 0.00895457, 0.00811491,
        0.00791202, 0.00831451, 0.00791826, 0.0078136 , 0.00779872,
        0.00803595, 0.00817561, 0.00942411, 0.00869961, 0.00761342,
        0.00781455, 0.00811186,

In [25]:
 grid_search.cv_results_.keys()
#grid_search.cv_results_['params']

dict_keys(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time', 'param_criterion', 'param_max_depth', 'param_min_samples_leaf', 'param_min_samples_split', 'param_random_state', 'params', 'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score', 'mean_test_score', 'std_test_score', 'rank_test_score'])

In [30]:
len(grid_search.cv_results_['param_criterion']=='gini') 

288

In [32]:
len(grid_search.cv_results_['params'])

288

In [33]:
grid_search.cv_results_['split0_test_score']

# p0 p1 p2 p3= teain p5=test
#288 params models 288 times developed 288 accyracy

array([0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82018349,
       0.82018349, 0.82018349, 0.82018349, 0.82018349, 0.82018349,
       0.82018349, 0.82018349, 0.82018349, 0.82018349, 0.82018349,
       0.82018349, 0.82018349, 0.82018349, 0.82018349, 0.82018349,
       0.82018349, 0.82018349, 0.82201835, 0.82201835, 0.82201835,
       0.82201835, 0.82201835, 0.82201835, 0.84587156, 0.84587156,
       0.84587156, 0.84587156, 0.84587156, 0.84587156, 0.84587156,
       0.84587156, 0.84587156, 0.84587156, 0.84587156, 0.84587156,
       0.84587156, 0.84587156, 0.84587156, 0.84587156, 0.84587156,
       0.84587156, 0.84770642, 0.84770642, 0.84770642, 0.84770642,
       0.84770642, 0.84770642, 0.83853211, 0.84220183, 0.83853

In [35]:
np.max(grid_search.cv_results_['split0_test_score'])

0.8899082568807339

In [42]:
param2_acc2=grid_search.cv_results_['split1_test_score'][0]
param3_acc3=grid_search.cv_results_['split2_test_score'][0]
param4_acc4=grid_search.cv_results_['split3_test_score'][0]
param5_acc5=grid_search.cv_results_['split4_test_score'][0]
print(param1_acc1,param2_acc2,param3_acc3,param4_acc4,param5_acc5)
param1_avg_acc=np.mean([param1_acc1,param2_acc2,param3_acc3,param4_acc4,param5_acc5])
param1_avg_acc
grid_search.cv_results_['mean_test_score'][0]

0.8275229357798165 0.8293577981651377 0.8146788990825689 0.8256880733944955 0.8106617647058824


0.8215818942255801

In [40]:
np.max(grid_search.cv_results_['mean_test_score'])

0.8649001618996222

In [41]:
grid_search.best_score_

0.8649001618996222

**np.argmax**

In [46]:
idx=np.argmax(grid_search.cv_results_['mean_test_score'])
grid_search.cv_results_['params'][idx]

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

In [47]:
grid_search.best_params_

{'criterion': 'entropy',
 'max_depth': 8,
 'min_samples_leaf': 4,
 'min_samples_split': 2,
 'random_state': 0}

In [50]:
grid_search.best_estimator_

DecisionTreeClassifier(criterion='entropy', max_depth=8, min_samples_leaf=4,
                       random_state=0)